synapse/lib/cmdr.py

```diff
@@ -9,36 +20,595 @@
+    '''
+    A client-side helper for connecting to a CLI server.
+
+    This acts as the client side pump for sending commands to a CLI server
+    and printing events from the CliServer. These events names are dynamically
+    generated, as sessions are created with the CliServer and as commands are
+    individually executed.
+
+    When the proxy is being initialized to a CliServer, it generates its own
+    proxy iden. This is used to set ``syn:cliserver:init:print:<iden>`` and
+    ``syn:cliserver:init:prompt:<iden>`` events, which the CLI server may use to
+    both send messages to the CliProxy for display to the user, and prompt
+    the CliProxy for additional input which is obtained through self.get_input().
+    This process is also used when addressing reconnect events.
+
+    After the session has been established with the CliServer, a session specific
+    iden is then used to monitor for syn:cliserver:<session_iden> events.  These
+    may be sent by the CliServer in order to send an arbitrary message to the
+    CliProxy to be displayed to the user.
+
+    When the user requests a specific command to be executed, a per-command iden
+    is generated, and the event syn:cliserver:outp:<cmdiden> is used to listen
+    for messages from the CliServer in response to the requested command. This
+    allows for a user to use ctrl+c to exit out of a potentially long running
+    command and no longer receive any messages for that commands execution.
+    '''
+    def __init__(self, prox, outp=None, timeout=None):
+        s_eventbus.EventBus.__init__(self)
+        self.disconnect_evt = threading.Event()
+
+        if outp is None:
+            outp = s_output.OutPut()
+        self.outp = outp
+
+        self.iden = s_common.guid()  # per-instance guid
+        self.cmdprompt = ''
+        self._quit_cmd = 'quit'  # Add set/get api?
+        self.prox = prox  # type: s_telepath.Proxy
+
+        self._connectToServer()
+
+        def _setDisconnectEvt(mesg):
+            already_set = self.disconnect_evt.is_set()
+            self.disconnect_evt.set()
+            # Only print this message and fire this once
+            if not already_set:
+                self.printf('Server disconnect detected.')
+                self.fire('syn:cliproxy:disconnect')
+
+        # Allow for per-cli messages to be sent to the CliProxy
+        self.prox.on('syn:cliserver:outp:%s' % self.session_iden, self._onPrintf)
+
+        # setup disconnect / reconnect handlers
+        self.prox.on('tele:sock:runsockfini', _setDisconnectEvt)
+        self.prox.on('tele:sock:init', self._onTeleSockInit)
+
+        self.cmdprompt = self.prox.getCmdPrompt()
+
+        self.onfini(self._cliProxFini)
+
+    def _onTeleSockInit(self, mesg):
+        '''
+        When the proxy object reconnects, attempt a reconnect to the CLI server.
+        '''
+        self._reconnect()
+        self.disconnect_evt.clear()
+
+    def _reconnect(self):
```

Comment:
Reconnect observation: If I stand up a server, then connect to it via CLI, then restart the server, then immediately run a query, the query seems to hang forever. If I cancel and rerun the query, it works. If I restart the server and wait a few seconds, the query also works.

Should reconnecting retry or raise an exception for the currently running query?

Reply:
Nevermind - This actually does work. I was connecting to the shared core instead of the shared CLI server. It seems unintuitive though that we have to stand up a separate server/object to use the CLI where we didn't have to before.